### PR TITLE
[SPARK-34889][SS] Introduce MergingSessionsIterator merging elements directly which belong to the same session

### PR DESCRIPTION
**Pull request by HeartSaVioR**
Introduction: this PR is a part of SPARK-10816 (`EventTime based sessionization (session window)`). Please refer #31937 to see the overall view of the code change. (Note that code diff could be diverged a bit.)

### What changes were proposed in this pull request?

This PR introduces MergingSessionsIterator, which enables to merge elements belong to the same session directly.

MergingSessionsIterator is a variant of SortAggregateIterator which merges the session windows based on the fact input rows are sorted by "group keys + the start time of session window". When merging windows, MergingSessionsIterator also applies aggregations on merged window, which eliminates the necessity on buffering inputs (which requires copying rows) and update the session spec for each input.

MergingSessionsIterator is quite performant compared to UpdatingSessionsIterator brought by SPARK-34888. Note that MergingSessionsIterator can only apply to the cases aggregation can be applied altogether, so there're still rooms for UpdatingSessionIterator to be used.

This issue also introduces MergingSessionsExec which is the physical node on leveraging MergingSessionsIterator to sort the input rows and aggregate rows according to the session windows.

### Why are the changes needed?

This part is a one of required on implementing SPARK-10816.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New test suite added.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
I forgot I cc.ed no one here. lol

cc. @viirya @xuanyuanking Please take a look. Thanks in advance!

**Comment by viirya** [comment]
Yea, I will review this this week.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+  val groupingWithoutSession: Seq[NamedExpression] =
+    groupingExpressions.diff(Seq(sessionExpression))
+  val groupingWithoutSessionAttributes: Seq[Attribute] = groupingWithoutSession.map(_.toAttribute)
+
+  /**
+   * Creates a new aggregation buffer and initializes buffer values
+   * for all aggregate functions.
+   */
+  private def newBuffer: InternalRow = {
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala:64]: Not a big deal and can be addressed later. Duplicated code with `SortBasedAggregationIterator`, I think we can avoid this by adding some abstractions.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+  private def getSessionStart(sessionStruct: UnsafeRow): Long = {
+    sessionStruct.getLong(0)
+  }
+
+  private def getSessionEnd(sessionStruct: UnsafeRow): Long = {
+    sessionStruct.getLong(1)
+  }
+
+  def updateSessionEnd(sessionStruct: UnsafeRow, sessionEnd: Long): Unit = {
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala:197]: only be used once in `expandEndOfCurrentSession`, maybe not necessary to be a function?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+    sessionStruct.setLong(1, sessionEnd)
+  }
+
+  private def expandEndOfCurrentSession(sessionEnd: Long): Unit = {
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala:201]: seems only used once in `processCurrentSortedGroup`? If so, maybe make it an inner function.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+    // We have not seen a new group. It means that there is no new row in the input
+    // iter. The current group is the last group of the iter.
+    if (!findNextPartition) {
+      sortedInputHasNewGroup = false
+    }
+  }
+
+  private def startNewSession(currentRow: InternalRow, groupingKey: UnsafeRow,
+                              sessionStruct: UnsafeRow): Unit = {
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala:183]: nit: indent

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+    }
+    sortedInputHasNewGroup
+  }
+
+  override final def next(): UnsafeRow = {
+    if (hasNext) {
+      // Process the current group.
+      processCurrentSortedGroup()
+      // Generate output row for the current group.
+
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala:223]: nit: extra empty line

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+      if (currentGroupingKey == groupingKey) {
+        if (sessionStart < getSessionStart(currentSession)) {
+          errorOnIterator = true
+          throw new IllegalStateException("Input iterator is not sorted based on session!")
+        } else if (sessionStart <= getSessionEnd(currentSession)) {
+          // expanding session length if needed
+          expandEndOfCurrentSession(sessionEnd)
+          processRow(sortBasedAggregationBuffer, currentRow)
+        } else {
+          // We find a new group.
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala:164]: To distinguish with line 169, maybe comment `we find a new session window for the same grouping key` here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+  initialize()
+
+  /** Processes rows in the current group. It will stop when it find a new group. */
+  protected def processCurrentSortedGroup(): Unit = {
+    currentGroupingKey = nextGroupingKey
+    currentSession = nextGroupingSession
+
+    // Now, we will start to find all rows belonging to this group.
+    // We create a variable to track if we see the next group.
+    var findNextPartition = false
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala:138]: A little bit confuse about the naming `findNextPartition`. Followed by the code, IIUC, the meaning might be `findNewWindow` or `findNextGroup`?

**Review by xuanyuanking — COMMENTED** [review]
Just some small comments. The logic is clear and majorly follows the SortAggregationIterator/Exec.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala (hunk 1 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala:64]: Unfortunately I don't remember clear as the original code is 2.5+ years ago, but I guess I tried to deduplicate with SortBasedAggregationIterator but gave up due to its complexity. If someone has brilliant idea to deduplicate then I'm happy to.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala (hunk 7 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala:138]: That was actually followed by SortBasedAggregationIterator, but findNextGroup should be clearer (even for SortBasedAggregationIterator). Let's change the name here for now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala (hunk 2 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala:197]: will inline.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala (hunk 3 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala:201]: Let's refrain `processCurrentSortedGroup` to deal with session struct directly. It'd be odd if we use functions to get the value but directly set it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsExec.scala [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+case class MergingSessionsExec(
+    requiredChildDistributionExpressions: Option[Seq[Expression]],
+    requiredChildDistributionOption: Option[Seq[Distribution]],
+    groupingExpressions: Seq[NamedExpression],
+    sessionExpression: NamedExpression,
+    aggregateExpressions: Seq[AggregateExpression],
+    aggregateAttributes: Seq[Attribute],
+    initialInputBufferOffset: Int,
+    resultExpressions: Seq[NamedExpression],
+    child: SparkPlan) extends UnaryExecNode {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsExec.scala:51]: Do we need to extend `BaseAggregateExec` like `SortAggregateExec`? `producedAttributes`, `aggregateBufferAttributes`look like the same.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala (hunk 1 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala:64]: We can do duplication later if needed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+  }
+
+  private val join = new JoinedRow
+
+  private val groupingKeyProj = GenerateUnsafeProjection.generate(groupingExpressions,
+    groupingWithoutSessionAttributes :+ sessionExpression.toAttribute)
+
+  private def generateGroupingKey(): UnsafeRow = {
+    val newRow = new SpecificInternalRow(Seq(sessionExpression.toAttribute).toStructType)
+    newRow.update(0, currentSession)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala:241]: newRow -> session or sessionStruct?

**Review by viirya — COMMENTED** [review]
Looks okay. By chance I just looked at `SortBasedAggregationIterator` today for other purpose. This looks pretty close to it so I don't have too many questions here.

Only one concern is the duplication which @xuanyuanking also pointed out before. But I think it is minor issue. We can consider do refactoring in the future.

**Comment by xuanyuanking** [comment]
Yea, agree on the duplication is a minor issue for now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala (hunk 3 above)
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala:199]: Make sense.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala (hunk 1 above)
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala:64]: Yea, agree.

**Review by xuanyuanking — APPROVED** [review]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsExec.scala (hunk 8 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsExec.scala:51]: Let's do it as it doesn't look to break the new tests.

**Comment by HeartSaVioR** [comment]
NOTE: I reran GA manually and somewhat the actual verification is gone. Please don't merge relying on GA build, and check the build via https://github.com/HeartSaVioR/spark/actions/runs/963446874 instead.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsExec.scala (hunk 8 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsExec.scala]: seems okay, the tests were passed.

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
Thanks @HeartSaVioR and @xuanyuanking! Merging to master.

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and merging!

**Pull request closed without merging**